### ISSUE TITLE
NegativeArraySizeException

### ISSUE THREAD

**Issue by andreak** [impossibl/pgjdbc-ng#166]
Using 0.5 driver with PG-9.3.5

When creating a PreparedStatement with LOTS of placeholders I get NegativeArraySizeException:

```
java.lang.NegativeArraySizeException
    at com.impossibl.postgres.protocol.v30.ProtocolImpl.receiveParameterDescriptions(ProtocolImpl.java:956)
    at com.impossibl.postgres.protocol.v30.ProtocolImpl.dispatch(ProtocolImpl.java:782)
    at com.impossibl.postgres.protocol.v30.MessageHandler.channelRead(MessageHandler.java:46)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:242)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:847)
    at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
    at java.lang.Thread.run(Thread.java:745)
```

paramCount in ProtocolImpl.receiveParameterDescriptions:

```
    short paramCount = buffer.readShort();

    TypeRef[] paramTypes = new TypeRef[paramCount];
```

returns -10198

This makes PGPreparedStatement.parseIfNeeded throw ClassCastException on line 217 here:

```
      catch (Exception e) {
        throw (SQLException) e;
      }
```

**Comment by jesperpedersen** [comment]
@andreak Could you retry with 0.6 ?

**Comment by jesperpedersen** [comment]
Ignore --- more coffee...

**Comment by jesperpedersen** [comment]
@andreak Do you have a test case for this ? Or the JDBC / SQL leading to this ?

**Comment by andreak** [comment]
Sure:

```
@RunWith(JUnit4.class)
public class NegativeArraySizeExceptionTest {
    private Connection conn;

    @Before
    public void before() throws Exception {
        conn = TestUtil.openDB();
        TestUtil.createTable(conn, "person", "id bigint primary key");
    }

    @After
    public void after() throws SQLException {
        TestUtil.dropTable(conn, "person");
        TestUtil.closeDB(conn);
    }

    @Test
    public void testLotsOfBindParams() throws SQLException {
        char fisk[] = new char[100000];
        for (int i = 0; i < fisk.length; i++) {
            fisk[i] = '?';
        }
        String csv = Arrays.toString(fisk).substring(1);
        csv = csv.substring(0, csv.length() - 1);
        String sql = String.format("SELECT 1 FROM person p WHERE p.id IN (%s)", csv);
        PreparedStatement ps = conn.prepareStatement(sql);
        for (int i = 1; i < fisk.length + 1; i++) {
            ps.setLong(i, i);
        }
        ps.executeQuery();
        ps.close();

    }

}
```

results in 

```
java.lang.ClassCastException: java.lang.NegativeArraySizeException cannot be cast to java.sql.SQLException
    at com.impossibl.postgres.jdbc.PGPreparedStatement.parseIfNeeded(PGPreparedStatement.java:195)
    at com.impossibl.postgres.jdbc.PGPreparedStatement.execute(PGPreparedStatement.java:234)
    at com.impossibl.postgres.jdbc.PGPreparedStatement.executeQuery(PGPreparedStatement.java:255)
    at com.impossibl.postgres.jdbc.PGPreparedStatement.executeQuery(PGPreparedStatement.java:92)
    at com.impossibl.postgres.jdbc.NegativeArraySizeExceptionTest.testLotsOfBindParams(NegativeArraySizeExceptionTest.java:43)
```

**Comment by kdubb** [comment]
Fixed in #263 although there is a limit of 65535 bound parameters. A proper SQLException is now reported.
